spharpy/spherical.py

```diff
@@ -1,104 +1,599 @@
+
+    @condon_shortley.setter
+    def condon_shortley(self, value):
+        if not isinstance(value, bool):
+            raise TypeError("condon_shortley must be a boolean value")
+        if value != self._condon_shortley:
+            self._recompute_basis = True
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._condon_shortley = value
+
+    @property
+    def weights(self):
+        return self._weights
+
+    @weights.setter
+    def weights(self, value):
+        if value is not None:
+            if not isinstance(value, (list, np.ndarray)):
+                raise TypeError("weights must be a list or numpy array")
+            if len(value) != self.coordinates.csize:
+                raise ValueError("weights must have the same length "
+                                 "as the number of coordinates")
+            if any(value < 0):
+                raise ValueError("weights must be non-negative")
+            if any(value == 0):
+                warnings.warn("Zero weights may lead to singular matrices")
+            if value != self._weights:
+                self._recompute_basis = True
+                self._recompute_basis_gradient = True
+                self._recompute_inverse = True
+                self._recompute_inverse_gradient = True
+        self._weights = value
+
+    @property
+    def n_max(self):
+        return self._n_max
+
+    @n_max.setter
+    def n_max(self, value):
+        if value < 0:
+            raise ValueError("n_max must be a positive integer")
+        if value % 1 != 0:
+            raise ValueError("n_max must be an integer value")
+        if value != self._n_max:
+            self._recompute_basis = True
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._n_max = int(value)  # Cast to int for safety
+
+    @property
+    def coordinates(self):
+        return self._coordinates
+
+    @coordinates.setter
+    def coordinates(self, value):
+        if not isinstance(value, pf.Coordinates):
+            raise TypeError("coordinates must be a pyfar.Coordinates "
+                            "object or spharpy.SamplingSphere object")
+        if value.cdim != 1:
+            raise ValueError("Coordinates must be 1D")
+            # TODO: Allow for 2D coordinates
+        if value != self._coordinates:
+            self._recompute_basis = True
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._coordinates = value
+
+    @property
+    def basis_type(self):
+        return self._basis_type
+
+    @basis_type.setter
+    def basis_type(self, value):
+        if value not in ["complex", "real"]:
+            raise ValueError("Invalid basis type, currently only "
```

Comment:
Maybe remove the currently? Which basis do we want to implement except real or complex?